grpc.go

```diff
@@ -0,0 +1,698 @@
+				exitCode = exitSignalOffset + int(waitStatus.Signal())
+				fieldLogger.WithField("exit-code", exitCode).Info("process was signaled")
+			} else {
+				exitCode = exitStatus
+				fieldLogger.WithField("exit-code", exitCode).Info("got wait exit code")
+			}
+		}
+
+	} else {
+		fieldLogger.Error("Process state is nil could not get process exit code")
+	}
+
+	return &pb.WaitProcessResponse{
+		Status: int32(exitCode),
+	}, nil
+}
+
+func (a *agentGRPC) RemoveContainer(ctx context.Context, req *pb.RemoveContainerRequest) (*gpb.Empty, error) {
+	ctr, err := a.pod.getContainer(req.ContainerId)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	a.pod.Lock()
+	defer a.pod.Unlock()
+	if err := ctr.removeContainer(); err != nil {
+		return emptyResp, err
+	}
+
+	delete(a.pod.containers, ctr.id)
+
+	return emptyResp, nil
+}
+
+func (a *agentGRPC) WriteStdin(ctx context.Context, req *pb.WriteStreamRequest) (*pb.WriteStreamResponse, error) {
+	proc, _, err := a.pod.getRunningProcess(req.ContainerId, req.ProcessId)
+	if err != nil {
+		return &pb.WriteStreamResponse{}, err
+	}
+
+	var file *os.File
+	if proc.termMaster != nil {
+		file = proc.termMaster
+	} else {
+		file = proc.stdin
+	}
+
+	n, err := file.Write(req.Data)
+	if err != nil {
+		return &pb.WriteStreamResponse{}, err
+	}
+
+	return &pb.WriteStreamResponse{
+		Len: uint32(n),
+	}, nil
+}
+
+func (a *agentGRPC) ReadStdout(ctx context.Context, req *pb.ReadStreamRequest) (*pb.ReadStreamResponse, error) {
+	data, err := a.pod.readStdio(req.ContainerId, req.ProcessId, int(req.Len), true)
+	if err != nil {
+		return &pb.ReadStreamResponse{}, err
+	}
+
+	return &pb.ReadStreamResponse{
+		Data: data,
+	}, nil
+}
+
+func (a *agentGRPC) ReadStderr(ctx context.Context, req *pb.ReadStreamRequest) (*pb.ReadStreamResponse, error) {
+	data, err := a.pod.readStdio(req.ContainerId, req.ProcessId, int(req.Len), false)
+	if err != nil {
+		return &pb.ReadStreamResponse{}, err
+	}
+
+	return &pb.ReadStreamResponse{
+		Data: data,
+	}, nil
+}
+
+func (a *agentGRPC) CloseStdin(ctx context.Context, req *pb.CloseStdinRequest) (*gpb.Empty, error) {
+	proc, _, err := a.pod.getRunningProcess(req.ContainerId, req.ProcessId)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	var file *os.File
+	if proc.termMaster != nil {
+		file = proc.termMaster
+	} else {
+		file = proc.stdin
+	}
+
+	if err := file.Close(); err != nil {
+		return emptyResp, err
+	}
+
+	return emptyResp, nil
+}
+
+func (a *agentGRPC) TtyWinResize(ctx context.Context, req *pb.TtyWinResizeRequest) (*gpb.Empty, error) {
+	proc, _, err := a.pod.getRunningProcess(req.ContainerId, req.ProcessId)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	if proc.termMaster == nil {
+		return emptyResp, fmt.Errorf("Terminal is not set, impossible to resize it")
+	}
+
+	fd := int(proc.termMaster.Fd())
+
+	// Get current terminal size.
+	winSize, err := unix.IoctlGetWinsize(fd, unix.TIOCGWINSZ)
+	if err != nil {
+		return emptyResp, err
+	}
```

Comment:
I don't think we need to get the winsz at first, for we are going to modify all the values.
```
       TIOCGWINSZ     struct winsize *argp
              Get window size.

       TIOCSWINSZ     const struct winsize *argp
              Set window size.

       The struct used by these ioctls is defined as

           struct winsize {
               unsigned short ws_row;
               unsigned short ws_col;
               unsigned short ws_xpixel;   /* unused */
               unsigned short ws_ypixel;   /* unused */
           };
```

Reply:
Well yes, if x and y position is unused I agree there is no need to get the window size before to set it.
@sameo any thoughts about this ?

Reply:
Agreed, you don't need to get winsize.

Reply:
Yeah I forgot to remove that, it's going to be fixed in next PR version.